src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -0,0 +1,124 @@
+using Azure.Data.AppConfiguration;
+using Microsoft.Extensions.Configuration.AzureAppConfiguration.Constants;
+using Microsoft.Extensions.Configuration.AzureAppConfiguration.Extensions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Extensions.Configuration.AzureAppConfiguration
+{
+    /// <summary>
+    /// A configuration client provider which maintains state of configuration clients and provides set of clients to use when requested.
+    /// </summary>
+    /// <remarks>
+    /// This class is not thread-safe. Since config provider does not allow multiple network requests at the same time,
+    /// there won't be multiple threads calling this client at the same time.
+    /// </remarks>
+    internal class ConfigurationClientManager : IConfigurationClientManager
+    {
+        private readonly IList<ConfigurationClientStatus> _clients;
+
+        public ConfigurationClientManager(string connectionString, ConfigurationClientOptions clientOptions)
+        {
+            if (string.IsNullOrEmpty(connectionString))
+            {
+                throw new ArgumentNullException(nameof(connectionString));
+            }
+
+            var endpoint = new Uri(ConnectionStringParser.Parse(connectionString, ConnectionStringParser.EndpointSection));
+            var configurationClientStatus = new ConfigurationClientStatus(endpoint, new ConfigurationClient(connectionString, clientOptions));
+            _clients = new List<ConfigurationClientStatus> { configurationClientStatus };
+        }
+
+        public ConfigurationClientManager(IEnumerable<Uri> endpoints, TokenCredential credential, ConfigurationClientOptions clientOptions)
+        {
+            if (endpoints == null || !endpoints.Any())
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            _clients = endpoints.Select(endpoint => new ConfigurationClientStatus(endpoint, new ConfigurationClient(endpoint, credential, clientOptions))).ToList();
+        }
+
+        public bool HasAvailableClients
+        {
+            get
+            {
+                var utcNow = DateTimeOffset.UtcNow;
+                return _clients.Any(client => client.BackoffEndTime <= utcNow);
+            }
+        }
+
+        /// <summary>
+        /// Internal constructor; Only used for unit testing.
+        /// </summary>
+        /// <param name="clients"></param>
+        internal ConfigurationClientManager(IList<ConfigurationClientStatus> clients)
+        {
+            _clients = clients;
+        }
+
+        public IEnumerable<ConfigurationClient> GetAvailableClients()
+        {
+            var utcNow = DateTimeOffset.UtcNow;
+            IEnumerable<ConfigurationClient> clients = _clients.Where(client => client.BackoffEndTime <= utcNow).Select(c => c.Client);
+
+            if (!clients.Any())
+            {
+                clients = _clients.Select(client => client.Client);
+            }
```

Comment:
I thought we decided to not return all clients anymore if all of them are in backoff state (yes, a behavior change). Did I remember it incorrectly?

The GetAvailableClients doesn't always agree with HasAvailableClients if we keep the current implementation. We need to think of a better naming.

Reply:
Correct, that's what we had decided earlier, but in offline discussion with Jimmy, we again revisited the decision to return all clients in case every client is in the backoff state and let the caller decide for themselves what to do when no clients are available (Caller needs to check using `HasAvailableClients`).

But I agree, the naming could be confusing, we need to think of a different name. How does `GetClients` sound to you?

Reply:
What I see in your current implementation is that the caller always check `HasAvailableClient`, so what's the point to return all clients?

Reply:
For the current use cases, there isn't a point. But it establishes a contract that it will give out at least one client when requested.
In future if we encounter a case where we wanted to make a network request regardless of client's backoff state, then the caller could just call `GetClients` without calling `HasAvailableClients` first.

Reply:
You eliminate the cache extension on failure cases and rely on clients being put in the backoff state, this function must NOT return any clients that are in the backoff state. Otherwise, your whole story won't hold.

I don't see the need of HasAvailableClients function. The callers should just call GetAvailableClients and do nothing if it's empty.

Reply:
I agree!